akka-stream/src/main/scala/akka/stream/impl/StreamLayout.scala

```diff
@@ -345,10 +348,32 @@ object StreamLayout {
       if (s != shape) throw new UnsupportedOperationException("cannot replace the shape of the EmptyModule")
       else this
 
-    override def compose(that: Module): Module = that
-
-    override def compose[A, B, C](that: Module, f: (A, B) ⇒ C): Module =
-      throw new UnsupportedOperationException("It is invalid to combine materialized value with EmptyModule")
+    override def compose(that: Module): Module = compose(that, scaladsl.Keep.left)
+
+    override def compose[A, B, C](that: Module, f: (A, B) ⇒ C): Module = {
+      if (f == scaladsl.Keep.right) {
+        that
+      } else if (f == scaladsl.Keep.left) {
```

Comment:
Hmm... So technically this is scala `Function`, which can not be the javadsl one, however if someone passed in javadsl.Keep.right then it'd be adapted won't it (on scala 2.12).

Not that that matters, and we don't need to check javadsl here now that I think about it it seems